Close #792: Rework Eglot's mode-line

nemethf opened the pull request:
I descended not converting the readme to org-mode and abandoned idea of adding a menu item to the mode-line that opens the local readme.   We can do it later, but it feels unnecessary because the documentation in info format will soon make it obsolete.

----

* eglot.el (eglot--mode-line-format): Capitalize tooltips; add
links to customizations and the readme on github; follow Emacs's
convention of "mouse-2 means help"; establish a convention of
"S-mouse-1 means customize"; move eglot-shutdown to C-mouse-3.

skangas commented:
Personally, I find this user interface to be a bit opaque, in particular because there are so many keys to remember (even if they are shown in the tooltip, IIUC).

How about adding instead a context menu similar to what `vc` does on `mouse-1`?

joaotavora commented:
@nemethf Let me just say that there can be links to the Info manual, too!  In SLY there is an info-manual-browsing command which is very useful.  So if that's the main reason to abandon this effort, maybe you could reconsider?

nemethf commented:
> How about adding instead a context menu similar to what `vc` does on `mouse-1`?

OK.  I've force pushed an update. However, I also followed the implementation of vc-menus and there are other (easier) ways to define menus.  Additionally, I haven't bothered to gray out unusable menu items (that are unusable because of missing server capabilities).  This can be done later.

The summary says it closes #792, but I'm not sure discussions can be closed automatically.  At any rate, I don't think the interface of "Discussions" is really suitable for discussing proposed features.

> @nemethf Let me just say that there can be links to the Info manual, too! In SLY there is an info-manual-browsing command which is very useful. So if that's the main reason to abandon this effort, maybe you could reconsider?

It seems I wasn't clear.  I won't convert README.md to README.org and add a menu item to show the local version, because this would be temporary: the info manual will be more useful to show than README.org would be.

skangas commented:
Felicián Németh ***@***.***> writes:

> How about adding instead a context menu similar to what vc does on mouse-1
> ?
>
> OK. I've force pushed an update. However, I also followed the
> implementation of vc-menus and there are other (easier) ways to define
> menus. Additionally, I haven't bothered to gray out unusable menu items
> (that are unusable because of missing server capabilities). This can be
> done later.
>
Sounds good, thanks. I'll try to take a look in the next couple of days.

> At any rate, I don't think the interface of "Discussions" is really
> suitable for discussing proposed features.
>
I tend to agree, and have discussed this privately with @joaotavora. It is
likely that we will be going back to using issues for tracking feature
requests, and of course it helps to hear your feedback in this regard.Message
ID: ***@***.***>

>

nemethf commented:
>  I haven't bothered to gray out unusable menu items (that are unusable because of missing server capabilities). This can be done later.

The most recent version grays out unusable menu items.  However, some items might not belong to the menu.  If that's the only problem with this PR, I propose merging it and refining the menu later.   Thanks.

skangas commented on eglot.el:
```diff
@@ -1700,6 +1700,97 @@ If it is activated, also signal textDocument/didOpen."
+         "Get possible code actions for the active region or the point"))
+    `("Eglot"
+      ;; Commands for getting information and customization.
+      [ "Read the documentation" eglot-read-documentation
```
This is very minor, but do we need space after `[` here? The indentation should be okay even without it for a vector.

nemethf replied:
I removed the space.

skangas commented on eglot.el:
```diff
@@ -1700,6 +1700,97 @@ If it is activated, also signal textDocument/didOpen."
+      [ "Read the documentation" eglot-read-documentation
+	:help "Read the on-line documentation"]
+      [ "Customize Eglot" eglot-customize
+	:help "Customize Eglot globally"]
```
Should customize and help be at the end perhaps? It seems like these would be less common actions than e.g. code actions.

nemethf replied:
The mode-line is at the bottom of the frame, so usually the bottom part of the menu is faster to access.  However, I don't think this menu should be used regularly.  It is for discoverability.  If seed is important, users should use key-bindings.  

So I haven't change the menu in this regard.  Nevertheless, if you rearrange the menu, that's fine with me.

skangas commented on eglot.el:
```diff
@@ -1723,18 +1814,28 @@ Uses THING, FACE, DEFS and PREPEND."
+         "eglot"
+         'face 'eglot-mode-line
+         'mouse-face 'mode-line-highlight
+         'help-echo "Eglot: an LSP client\nmouse-1: Display minor mode menu"
```
How about instead `"Eglot: Emacs LSP client"`?

nemethf replied:
The current text comes form the readme.  I think the Emacs here part is unnecessary because what else could it be.  Anyway, if you like to change it, that's fine with me.

skangas reviewed (commented):
Just some minor nit-picks.

skangas commented:
@nemethf BTW, could we have an entry in `NEWS.md` for this as well please?

nemethf commented:
> @nemethf BTW, could we have an entry in `NEWS.md` for this as well please?

I added a news entry and addressed your code-review comments.  If there are some minor issues left in the code, can you, please, fix them the way you like it.  Thanks.

joaotavora commented:
I tweaked this somewhat liberally (but not extensively) and merged it.  It's pretty useful so no point in letting it linger here. @nemethf let me know if you have more of these high-value PRs lying around :-)

nemethf commented:
> I tweaked this somewhat liberally (but not extensively) and merged it.

Your preferred style seems to differ from Stefan's, but I'm happy this has been merged.

> It's pretty useful so no point in letting it linger here. @nemethf let me know if you have more of these high-value PRs lying around :-)

There are some PRs that are easy to review and merge / reject, but none of them has "high-value":
- #905, #838, 
- #854, #818
